### PR TITLE
[DON'T MERGE][Discussion] Prototype for Stochastic Processes and Markov Chains

### PR DESCRIPTION
**Pull request by czgdp1807**
#### References

[1] [Wikipedia of Markov Chains](https://en.wikipedia.org/wiki/Markov_chain)
[2] Project discussions at https://gitter.im/sympy/stats
[1] [Protoype in GSoC Proposal](https://docs.google.com/document/d/1oIeaROiJyglpbris7X1uZPRE5ZeO0pD1ygCFhBIeATI/edit#heading=h.en3qajombcaj)

#### Brief description of what is fixed or changed
Prototype for the purpose of discussion has been added. 
The proposed architecture of stochastic processes is as follows:
1. `StochasticProcess` - The base class for computations of properties related to various types of stochastic processes. It is capable of generating `StochasticRandomSymbol` using its `value` method.
2. `StochasticRandomSymbol` - The class which inherits `RandomSymbol` and redefines `__getitem__`  method for allowing the accessibility of infinite times. 
3. `AbstractStochasticProcess` - The base class for various types of stochastic processes like Markov chains, Random Walks, etc.
4. `MarkovChain` - The class for defining specific properties and extra properties related to `MarkovChain`.
5. `MarkovChainRV` - The interface function for creating random variables of `MarkovChain`.
5. `stochastic_rv` - The interface function to create `StochasticRandomSymbol`.

#### Other comments
Since, only prototypes is added, no tests has been added. The aim of this PR is to discuss upon the architecture rather than the specific details.
**PLEASE DO NOT MERGE OR CLOSE THIS PULL REQUEST**

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by czgdp1807** [comment]
@Upabjojr @sidhantnagpal Please comment your reviews on the structure of the prototype and any improvements that can be made.

**Comment by Upabjojr** [comment]
Seems like a nice proposal.

However I keep recommending starting with the simplest implementation possible on some example problems. The simpler you are able to design the classes, the better it is.

**Comment by Upabjojr** [comment]
I'm not even sure we need to have a specific `StochasticRandomSymbol`.

Maybe `IndexedRandomSymbol` could be enough. Or maybe just extending `RandomSymbol` to support `__getitem__` could provide a solution.

Also, stochastic processes may be seen as infinite-dimensional joint probability vectors of random variables. Random matrices may be seen as two dimensional arrays of random matrices. Could we use the same object for both cases?

File: sympy/stats/stochastic_process.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    @property
+    def process(self):
+        return self.args[1]
+
+    @property
+    def value(self):
+        return StochasticRandomSymbol(self.symbol, self)
+
+    def joint_dist(self, n):
+        return NotImplementedError()
```
Comment (supreet11agrawal) [inline sympy/stats/stochastic_process.py]: You need to raise an exception here. So, change it to `raise NotImplementedError()`.

File: sympy/stats/stochastic_process.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+    def first_passage_time(self, state):
+        return NotImplementedError()
```
Comment (supreet11agrawal) [inline sympy/stats/stochastic_process.py]: Same here.

File: sympy/stats/stochastic_process.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+    def probability(self, initial_state):
+        return NotImplementedError()
```
Comment (supreet11agrawal) [inline sympy/stats/stochastic_process.py]: Same here.

**Review by supreet11agrawal — CHANGES_REQUESTED** [review]

File: sympy/stats/stochastic_process.py (hunk 1 above)
Comment (czgdp1807) [reply sympy/stats/stochastic_process.py]: Isn't it already `raise NotImplementedError()`?

Comment (czgdp1807) [reply sympy/stats/stochastic_process.py]: Well that's just a prototype, not a big deal.

**Comment by supreet11agrawal** [comment]
One question, should we really make `states` necessary? I mean, one might just need properties related to transition matrix.

**Comment by czgdp1807** [comment]
@supreet11agrawal According to the formal definition, `state_space` should be there in the properties.  There is nothing like `states` in the prototype.

**Comment by Upabjojr** [comment]
What about infinite dimensional states (possibly uncountable) in Markov chains?

**Comment by czgdp1807** [comment]
As this PR has been referenced in #16981. Closing.

**Pull request closed without merging**